mdsobjects/cpp/mdstreeobjects.cpp

```diff
@@ -1661,7 +1681,337 @@ StringArray *TreeNodeArray::getUsage()
+    return as.string;
+}
+
+char TreeNodeThinClient::getNciChar(int itm)
+{
+    char expr[64];
+    sprintf(expr, "GETNCI(%d,%s)", nid, convertNciItm(itm));
+    Data *retData = connection->get(expr);
+     if(!retData)
+	throw MdsException("Error in Remote evaluation of getnci");
+    char retChar = retData->getByte();
+    MDSplus::deleteData(retData);
+    return retChar;
+}
+   
+int TreeNodeThinClient::getNciInt(int itm)
+{
+    char expr[64];
+    sprintf(expr, "GETNCI(%d,%s)", nid, convertNciItm(itm));
+    Data *retData = connection->get(expr);
+    if(!retData)
+	throw MdsException("Error in Remote evaluation of getnci");
+    int retInt = retData->getInt();
+    MDSplus::deleteData(retData);
+    return retInt;
+}
+
+int64_t TreeNodeThinClient::getNciInt64(int itm)
+{
+    char expr[64];
+    sprintf(expr, "GETNCI(%d,%s)", nid, convertNciItm(itm));
+    Data *retData = connection->get(expr);
+    if(!retData)
+	throw MdsException("Error in Remote evaluation of getnci");
+    int64_t retLong = retData->getLong();
+    MDSplus::deleteData(retData);
+    return retLong;
+}
+
+char *TreeNodeThinClient::getPath()
+{
+    char expr[64];
+    sprintf(expr, "GETNCI(%d,\'PATH\')", nid);
+    Data *retData = connection->get(expr);
+    if(!retData)
+	throw MdsException("Error in Remote evaluation of getnci(path)");
+    char *retPath = retData->getString();
+    MDSplus::deleteData(retData);
+    return retPath;
+}
+    
+EXPORT Data *TreeNodeThinClient::getData()
+{
+    char expr[64];
+    sprintf(expr, "GETNCI(%d,\'RECORD\')", nid);
+    Data *retData = connection->get(expr);
+    if(!retData)
+	throw MdsException("Error in Remote evaluation of getnci(record)");
+    return retData;
+}
+  
+EXPORT void TreeNodeThinClient::putData(Data *data)
+{
+    const char *path = (const char *)getPath();
+    Data *args[1];
+    args[0] = data->data();
+    connection->put(path, (char *)"$", args, 1);
+    MDSplus::deleteData(args[0]);
+    delete [] path;
+  
+}
+
+EXPORT void TreeNodeThinClient::deleteData()
+{
+    char *path = getFullPath();
+    Data *args[1];
+    args[0] = 0;
+    connection->put(path, (char *)"*", args, 0);
+    delete [] path;
+}
+
+EXPORT bool TreeNodeThinClient::isOn()
+{
+    char expr[64];
+    sprintf(expr, "GETNCI(%d,\'STATE\')", nid);
+    Data *retData = connection->get(expr);
+    if(!retData)
+	throw MdsException("Error in Remote evaluation of getnci(state)");
+    
+    char onData = retData->getByte();
+    MDSplus::deleteData(retData);
+    return (onData)?false:true;
+}
+
+EXPORT void TreeNodeThinClient::setOn(bool on)
+{
+    char expr[64];
+    if(on)
+      sprintf(expr, "TreeTurnOn(%d)", nid);
+    else
+      sprintf(expr, "TreeTurnOff(%d)", nid);
+    Data *retData = connection->get(expr);
+    if(retData)
+	MDSplus::deleteData(retData);
+}
+
+EXPORT void TreeNodeThinClient::beginSegment(Data *start, Data *end, Data *time, Array *initialData)
+{
+    char expr[256];
+    Data *args[4];
```

Comment:
I believe I had a comment to do this inline:

Data * args[4] = {
start->data(),
 end->data(),
 time->data(),
 initialData->data()
};

This gets you basic compile checking of array bounds

And of course, making it be an array of AutoData's means that you can remove the for loop at the end (please do this)